### ISSUE TITLE
Issue with date display when browser's timezone is different from reporting timezone

### ISSUE THREAD

**Issue by HelenMertes** [metabase/metabase#5246]
Hi,
I am having issues displaying the data axis for bar charts based on SQL queries grouped by month properly. Using "ordinal" it shows the entire date incl time of the 1st day of the respective month, which looks awful and overlaps. With "timeseries" if correctly shows the label "June 2016", but not below the correct data point, i.e. below July (ad the label "July 2017" below the August data point etc). Any idea how to resolve this so that the chart can be displayed properly?
Thanks!

- Your browser and the version: Version 58.0.3029.110 (64-bit)
- Your operating system: Windows 10
- Your databases: BigQuery
- Metabase version: 24.1

**Comment by salsakran** [comment]
I'd guess this is an incorrect timezone conversion happening somewhere. 

Can you share the following (taken from http://www.metabase.com/troubleshooting/timezones) - 

What is the time zone of the data you think is being displayed improperly? (I.e., in the database itself.)
Are you using an explicit time zone setting on each timestamp, or are the timestamps being stored without a timestamp? (E.g., Dec 1, 2019 00:00:00Z00 is an explicitly timestamped value, but Dec 1, 2019 has an implied time zone.)
What time zone is the database server set to?
What time zone is the server that is running Metabase set to?
What is your Reporting Time zone setting?
What is your browser time zone setting?

**Comment by gilhs** [comment]
Not the OP, but also started experiencing this (seems to start with 0.24?), so will give my input - this also happens when visualizing SQL results, and specifically when using **dates** [i.e. date(<some_timestamp_filed>) which returns 'YYYY-MM-DD']. So this isn't just limited to timestamps.

![image](https://cloud.githubusercontent.com/assets/22723398/26776165/c5067cc2-49e0-11e7-896e-e883b8667fdb.png)

**Comment by gseva** [comment]
We are experiencing the same problem. I will add our case here to not open another issue. 

Seems like this happens when we change the timezone setting (via `/admin/settings/general`).

This is what we see with timezone setting set to `Brazil/East`:

<img width="737" alt="screen shot 2017-06-26 at 14 11 59" src="https://user-images.githubusercontent.com/2244417/27551494-03f5eb9e-5a7a-11e7-9ed5-a066706e2ad2.png">

That's the result with timezone set to `Database default` (UTC):

<img width="762" alt="screen shot 2017-06-26 at 14 12 32" src="https://user-images.githubusercontent.com/2244417/27551596-6528797c-5a7a-11e7-82d0-77fa653d352f.png">

The second chart shows correct data, in the first one all the counts are moved 1 day.

We are using Metabase 23.1, connecting to postgesql database. *Created at* field in our database is saved as `bigint`, and is defined as `Unix timestamp (seconds)` in Metabase's data model.

Hope it's helpful to fix this issue.

**Comment by ebalgava** [comment]
I'm having same issue with shifted date labels in charts:
https://bi.ekosystem.slovensko.digital/public/dashboard/86ce303c-05bf-4947-bb39-d2a958208492
@salsakran anounced 'this should be fixed in 0.26' but my issue still persists within v0.26.1:
https://github.com/metabase/metabase/issues/3642#issuecomment-332653232

**Comment by hennumjr** [comment]
Also having the same issue with version 0.26.1. The tool tip is correct, but axis tick labels are off by one.
![wrong-axis-tick-issue](https://user-images.githubusercontent.com/754494/31661715-fd6d59ea-b33b-11e7-9538-aa42ee109974.png)

**Comment by mnzt** [comment]
![image](https://user-images.githubusercontent.com/10098811/31941985-b2b420f2-b8bb-11e7-8efe-21ad7ceae734.png)

Experiencing the same on 0.26.1.

**Comment by mazameli** [comment]
One place to start for everyone who's experiencing this is to first go through the timezone troubleshooting doc here: https://github.com/metabase/metabase/blob/master/docs/troubleshooting-guide/timezones.md

**Comment by ebalgava** [comment]
@mazameli this issue definitely isn't timezone related, as the same chart shows correct result in value label, just X-axis label is shifted

**Comment by salsakran** [comment]
@ebalgava the bar charts at that link look like they have matched axis ticks and tool tips. 

Can anyone reproduce this with the sample dataset on 0.26.X? 

We fixed a number of timezone issues in the most recent release, but it's entirely possible we missed some corner cases. Can you share more information about the underlying query, database timezone, your browser timezone, etc, and help us reproduce your situation?

**Comment by salsakran** [comment]
@ebalgava fyi, there was in fact a class of timezone bugs related to situations where tooltips (or other representations) weren't correctly mapping to a timezone

**Comment by ebalgava** [comment]
@salsakran my example:
https://bi.ekosystem.slovensko.digital/public/dashboard/86ce303c-05bf-4947-bb39-d2a958208492

**Comment by salsakran** [comment]
This is what I see

![screen shot 2017-10-24 at 10 03 15 am](https://user-images.githubusercontent.com/107439/31957136-99bfa22e-b8a2-11e7-9514-e7f87aee6b11.png)

Can you point me to where in that dashboard you're seeing a discrepancy?

**Comment by ebalgava** [comment]
![diff](https://user-images.githubusercontent.com/29505976/31957313-ac12d260-b8ee-11e7-8b28-f29dd2157b03.png)

**Comment by calmyournerves** [comment]
Is this related to the browser time zone? It's wrong for me as well (CEST).

![image](https://user-images.githubusercontent.com/480921/31957332-b8d70d36-b8ee-11e7-82ec-8e7ee1086527.png)

**Comment by mnzt** [comment]
@salsakran Can you try casting to a DATE type? When grouping a timestamp by day (via casting to date) it appears Metabase interprets this as "Monday, October 23rd, 2017 12:00AM" in the tooltip, but appears as October 22nd on the X axis.

I'm guessing this is an issue surround the "12:00AM" bit.

**Comment by restyler** [comment]
I am having the same issue where x-axis label  is wrong, tooltip label is correct.
<img width="664" alt="screenshot at dec 30 19-40-46" src="https://user-images.githubusercontent.com/775507/34455479-5cc7a34a-ed99-11e7-945f-9e353267f902.png">

All the data is stored in UTC timestamp in mysql database.

**Comment by achesnais** [comment]
I've had a similar issue and it appears to come down to a discrepancy between the Reporting timezone setting and the browser time zone.

If I set my Reporting timezone to UTC, and leave my browser in UTC+2, I get the above issue

If I set my Reporting timezone to UTC+2 the discrepancy disappears.

If I set my Reporting timezone to UTC, and force my browser to use UTC, the issue disappears

If I set my Reporting timezone to UTC+2 and force my browser to use UTC, I don't get an issue on the chart, but I if I explore the data I find datums from the next month included in the current month.

Based on the above, it seems the chart uses the browser timezone to render the dates, but the value of the dates themselves is displayed using the reporting time zone? And I suppose the value shown when hovering over a bar is that of the earliest datum in the group? This would explain all cases above.

**Comment by TeamDKS** [comment]
@salsakran is there any news about this issue? because we have the same problem in our company. Possible workaround for others: if the order in the DB is correct it will work when you use the ordinal x-axis scale in the axis settings on the question.

**Comment by tlrobinson** [comment]
@HelenMertes @gilhs @gseva @ebalgava @calmyournerves @mnzt @restyler @achesnais @TeamDKS I believe we have a fix for this in #7247. Please give it a try and let me know if it does or doesn't fix this for you (and hopefully doesn't make things worse...)

**Comment by SetSails** [comment]
@tlrobinson any indication on when this will be merged into a stable version? I'm actually experiencing #3642 since over a year now, tried all sorts of timezone settings and am brutally hoping for this issue to be finally fixed.

Thanks & Best Regards
F

**Comment by tlrobinson** [comment]
@SetSails If you could pull down the branch in PR #7247 and validate that it fixes the issue for you that would be helpful.

**Comment by MarkRx** [comment]
@tlrobinson I can confirm that #7247 fixes the following use case:

```
select 
    start_month "Month",
    status,
    count(*) as "Deployments"
    from (
    select 
        xldt_task_id,
        to_timestamp(extract(month from xldt_start_dt) || '-' || extract(year from xldt_start_dt) , 'MM-YYYY') as start_month,
        xldt_duration,
        (case
            when xldt_state = 'CANCELLED' then 'ABORTED' 
            when xldt_failure_count > 0 then 'RETRIED'
            when xlddt_type = 'ROLLBACK' then 'ROLLBACK'
            when xldt_state = 'DONE' then 'SUCCESS'
        end) as status
    from mbase.deployments
    . . .
)
where start_month > sysdate - 365
group by start_month, status
```

Before (x-axis label is wrong):
![before](https://user-images.githubusercontent.com/17504314/46020951-8949f200-c0ad-11e8-81d5-a6b96b1dabea.PNG)

After:
![after](https://user-images.githubusercontent.com/17504314/46020954-8bac4c00-c0ad-11e8-910e-711ad65e4fc8.PNG)

**Comment by TrineFN** [comment]
Experiencing the same on 0.31.2

**Comment by azhestov** [comment]
Experiencing the same on 0.32.0

**Comment by antoine-lizee** [comment]
Still seeing this in **v0.32.5**.

From my quick understanding:

**My setup**:
- DB in UTC
- MB reporting tz = UTC
- timestamps without timezone
- **local timezone (browser) different from reporting tz** (CET in my case)

**What it's not:** the usual "data picker filter gets wrong data". I see the correct information in the table view.

![image](https://user-images.githubusercontent.com/2957716/58105976-a5277880-7be7-11e9-951c-29a943ae0c28.png)


**What it is:** Problem with how timezone is handled by the frontend, when creating graphs. Here is the exact same query after simply changing the Vizualisation in "Bar" mode.

![image](https://user-images.githubusercontent.com/2957716/58106078-ce480900-7be7-11e9-9e9d-efd68063ce3e.png)

**Impact:** Pretty big - this behavior can be seen in any chart with temporal data, even if created with a "native" query. Honestly this bug alone is making us consider switching reporting platform, because it's everywhere - we literally have a section about this caveat in our metabase onboarding material. The only way around we found is to actually format the date string ourselves.

**Fix**: It _seems_ that #7247 was meant to fix those bugs, but it's now stale. Is there anyway that we can revive it?

**Comment by markpaygo** [comment]
Is there going to be a fix for this issue soon? We are actively looking to switch to other data analytics platform because we can't display temporal data correctly

**Comment by antoine-lizee** [comment]
Hey there, I'm still seeing this with 0.33RC. 

The bug could be called: **local frontend timezone is breaking all charts that depend on dates**.
See my detailed report right above. It's been going on for years now.

It's really a **major bug** for any team that is not tiny (so with people on different timezones), and I don't understand why it's only P3. Can we revisit the priority on this one? Would it be helpful to open a new issue to better scope the problem?

**Comment by antoine-lizee** [comment]
Just a heads up here as we see some activity on this issue.

@mazameli @tlrobinson @paulrosenzweig What's the current ETA for a fix here?

**Comment by paulrosenzweig** [comment]
@antoine-lizee this should be fixed in the next release. There's [a PR up](https://github.com/metabase/metabase/pull/11111) if you want to check that it fixes the problem on your data.

**Comment by mazameli** [comment]
@paulrosenzweig, should this issue have been marked with the `0.33.5` milestone? I.e. it looks like #11111 was included in 0.33.5, but didn't get included in the release notes because this issue wasn't marked and closed. Is my read correct?

**Comment by paulrosenzweig** [comment]
@mazameli whoops my bad! Yes this went out with 0.33.5.

**Comment by paulrosenzweig** [comment]
Fixed in #11111
